src/agent/discovery_proxy.cpp

```diff
@@ -0,0 +1,369 @@
+        break;
+    default:
+        ExitNow(error = OTBR_ERROR_NOT_IMPLEMENTED);
+    }
+
+    it = std::find_if(mSubscriptions.begin(), mSubscriptions.end(), [&](const MdnsSubscription &aSubscription) {
+        return aSubscription.Matches(instanceName, serviceName, hostName, domain);
+    });
+
+    VerifyOrExit(it != mSubscriptions.end(), error = OTBR_ERROR_NOT_FOUND);
+
+    {
+        MdnsSubscription &subscription = *it;
+
+        subscription.mSubscriptionCount--;
+        assert(subscription.mSubscriptionCount >= 0);
+
+        if (subscription.mSubscriptionCount == 0)
+        {
+            mSubscriptions.erase(it);
+        }
+
+        otbrLog(OTBR_LOG_DEBUG, "[dnssd discovery proxy] service subscriptions: %sx%d", it->ToString().c_str(),
+                it->mSubscriptionCount);
+
+        if (GetServiceSubscriptionCount(instanceName, serviceName) == 0)
+        {
+            if (subscription.mHostName.empty())
+            {
+                mMdnsPublisher.UnsubscribeService(serviceName, instanceName);
+            }
+            else
+            {
+                mMdnsPublisher.UnsubscribeHost(hostName);
+            }
+        }
+    }
+exit:
+    if (error != OTBR_ERROR_NONE)
+    {
+        otbrLog(OTBR_LOG_WARNING, "[dnssd discovery proxy] failed to unsubscribe %s.%s: %s",
+                instanceName.empty() ? "ANY" : instanceName.c_str(), aFullName, otbrErrorString(error));
+    }
+}
+
+void DiscoveryProxy::OnServiceDiscovered(const std::string &                            aType,
+                                         const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo)
+{
+    otDnssdServiceInstanceInfo instanceInfo;
+
+    otbrLog(OTBR_LOG_INFO,
+            "[dnssd discovery proxy] service discovered: %s, instance %s hostname %s address %s:%d priority %d "
+            "weight %d",
+            aType.c_str(), aInstanceInfo.mName.c_str(), aInstanceInfo.mHostName.c_str(),
+            aInstanceInfo.mAddress.ToString().c_str(), aInstanceInfo.mPort, aInstanceInfo.mPriority,
+            aInstanceInfo.mWeight);
+
+    CheckServiceNameSanity(aType);
+    CheckHostnameSanity(aInstanceInfo.mHostName);
+
+    instanceInfo.mAddress   = *reinterpret_cast<const otIp6Address *>(&aInstanceInfo.mAddress);
+    instanceInfo.mPort      = aInstanceInfo.mPort;
+    instanceInfo.mPriority  = aInstanceInfo.mPriority;
+    instanceInfo.mWeight    = aInstanceInfo.mWeight;
+    instanceInfo.mTxtLength = static_cast<uint16_t>(aInstanceInfo.mTxtData.size());
+    instanceInfo.mTxtData   = aInstanceInfo.mTxtData.data();
+    instanceInfo.mTtl       = std::min(aInstanceInfo.mTtl, static_cast<uint32_t>(kServiceTtlCapLimit));
+
+    std::for_each(mSubscriptions.begin(), mSubscriptions.end(), [&](const MdnsSubscription &aSubscription) {
+        if (aSubscription.MatchesServiceInstance(aType, aInstanceInfo.mName))
+        {
+            std::string serviceFullName  = aType + "." + aSubscription.mDomain;
```

Comment:
Do we want to translate domain for service names as well?

Reply:
I think this statement is already doing it.